cpp/src/arrow/compute/expression-test.cc

```diff
@@ -0,0 +1,154 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#include "arrow/status.h"
+#include "arrow/table.h"
+#include "arrow/testing/gtest_common.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/type.h"
+
+#include "arrow/compute/expression.h"
+#include "arrow/compute/logical_type.h"
+#include "arrow/compute/operation.h"
+
+namespace arrow {
+namespace compute {
+
+// A placeholder operator implementation to use for testing various Expr
+// behavior
+class DummyOp : public Operation {
+ public:
+  Status ToExpr(std::shared_ptr<Expr>* out) const override {
+    return Status::NotImplemented("NYI");
+  }
+};
+
+TEST(TestLogicalType, NonNestedToString) {
+  std::vector<std::pair<std::shared_ptr<LogicalType>, std::string>> cases = {
+      {type::any(), "Any"},
+      {type::null(), "Null"},
+      {type::boolean(), "Bool"},
+      {type::number(), "Number"},
+      {type::floating(), "Floating"},
+      {type::integer(), "Integer"},
+      {type::signed_integer(), "SignedInteger"},
+      {type::unsigned_integer(), "UnsignedInteger"},
+      {type::int8(), "Int8"},
+      {type::int16(), "Int16"},
+      {type::int32(), "Int32"},
+      {type::int64(), "Int64"},
+      {type::uint8(), "UInt8"},
+      {type::uint16(), "UInt16"},
+      {type::uint32(), "UInt32"},
+      {type::uint64(), "UInt64"},
+      {type::float_(), "Float"},
```

Comment:
can we at least do float_t or maybe float type, for some reason the hanging underscore really bugs me.